File: InkingWorkaround/InkingWorkaround/InkBitmapRenderer.cpp
#include "pch.h"
#include "InkBitmapRenderer.h"

#include <Microsoft.Graphics.Canvas.native.h>
#include <d2d1_1.h>

using namespace InkingWorkaround;

using namespace Microsoft::Graphics::Canvas;
using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::Graphics::DirectX;
using namespace Windows::Graphics::DirectX::Direct3D11;
using namespace Windows::Graphics::Display;
using namespace Windows::Graphics::Imaging;
using namespace Windows::UI;
using namespace Windows::UI::Input::Inking;

InkBitmapRenderer::InkBitmapRenderer() : InkBitmapRenderer(true)
{
}


InkBitmapRenderer::InkBitmapRenderer(bool hardwareAccelerated)
	: InkBitmapRenderer(
		ref new CanvasDevice(
			CanvasDebugLevel::None,
			hardwareAccelerated ? CanvasHardwareAcceleration::Auto : CanvasHardwareAcceleration::Off
			),
		true
		)
{
}

InkBitmapRenderer::~InkBitmapRenderer()
{
}

InkBitmapRenderer^ InkBitmapRenderer::CreateWithDirect3D11Device(IDirect3DDevice^ sharedDevice)
{
	auto canvasDevice = CanvasDevice::CreateFromDirect3D11Device(
		sharedDevice,
		CanvasDebugLevel::None
		);
	return ref new InkBitmapRenderer(canvasDevice, false);
}

InkBitmapRenderer::InkBitmapRenderer(CanvasDevice^ canvasDevice, bool ownsDevice)
	: m_canvasDevice(canvasDevice), m_ownsDevice(ownsDevice)
{
	if (m_ownsDevice)
	{
		m_canvasDevice->DeviceLost += ref new TypedEventHandler<CanvasDevice^, Platform::Object^>(
			this,
			&InkBitmapRenderer::OnDeviceLost
			);
	}

	CoCreateInstance(
		__uuidof(InkD2DRenderer),
		nullptr,
		CLSCTX_INPROC_SERVER,
		IID_PPV_ARGS(&m_inkRenderer)
		);
}

void InkBitmapRenderer::UpdateDevice(IDirect3DDevice^ sharedDevice)
{
	m_canvasDevice = CanvasDevice::CreateFromDirect3D11Device(sharedDevice, CanvasDebugLevel::None);
	m_ownsDevice = false;
}

void InkBitmapRenderer::OnDeviceLost(CanvasDevice^ sender, Object^ args)
{
	if (m_ownsDevice && Object::ReferenceEquals(sender, m_canvasDevice))
	{
		RecreateDevice();
	}
}

void InkBitmapRenderer::RecreateDevice()
{
	m_canvasDevice = ref new CanvasDevice(
		CanvasDebugLevel::None,
		m_canvasDevice->HardwareAcceleration
		);
	m_canvasDevice->DeviceLost += ref new TypedEventHandler<CanvasDevice^, Platform::Object^>(
		this,
		&InkBitmapRenderer::OnDeviceLost
		);
}

IAsyncOperation<SoftwareBitmap^>^ InkBitmapRenderer::RenderStrokesAsync(
	IIterable<InkStroke^>^ inkStrokes,
	double width,
	double height
	)
{
	auto dpi = DisplayInformation::GetForCurrentView()->LogicalDpi;
	try
	{
		auto renderTarget = ref new CanvasRenderTarget(
			m_canvasDevice,
			static_cast<float>(width),
			static_cast<float>(height),
			dpi
			);
		{
			auto drawingSession = renderTarget->CreateDrawingSession();
			auto deviceContext = GetWrappedResource<ID2D1DeviceContext>(drawingSession);

			deviceContext->Clear();
			auto hr = m_inkRenderer->Draw(
				deviceContext.Get(),
				reinterpret_cast<IUnknown*>(inkStrokes),
				false
				);

			if (FAILED(hr))
			{
				throw ref new Platform::Exception(hr);
			}
		}

		return SoftwareBitmap::CreateCopyFromSurfaceAsync(renderTarget);
	}
	catch (Exception^ e)
	{
		if (m_canvasDevice->IsDeviceLost(e->HResult) && m_ownsDevice)
		{
			m_canvasDevice->RaiseDeviceLost();

			// Return null on device lost
			return concurrency::create_async([]() -> SoftwareBitmap^
			{
				return nullptr;
			});
		}
		else
		{
			// If the client provided their own device, they should catch the device lost HResult
			throw;
		}
	}
}

File: Common/InkBitmapRenderer.cs
using Microsoft.Graphics.Canvas;
using Microsoft.Graphics.Canvas.Geometry;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;
using Windows.Graphics.Display;
using Windows.Graphics.Imaging;
using Windows.UI.Input.Inking;
using System.Numerics;
using Windows.UI;
using Windows.UI.Xaml.Media;
using Microsoft.Graphics.Canvas.UI.Xaml;
using Windows.UI.Xaml.Media.Imaging;
using Windows.Foundation;

namespace InkRendering
{

    public class InkBitmapRenderer
    {
        private CanvasDevice _canvasDevice;

        public InkBitmapRenderer() : this(false)
        {
        }

        public InkBitmapRenderer(bool forceSoftwareRenderer)
        {
            _canvasDevice = CanvasDevice.GetSharedDevice(forceSoftwareRenderer);
            _canvasDevice.DeviceLost += HandleDeviceLost;
        }

        
        public  async Task<CanvasImageSource> RenderToImageSourceAsync(
            RenderTargetBitmap backgroundBitmap,
            IReadOnlyList<InkStroke> inkStrokes, IReadOnlyList<InkStroke> highlightStrokes, double width, double height)
        {

            var dpi = DisplayInformation.GetForCurrentView().LogicalDpi;
            var imageSource = new CanvasImageSource(_canvasDevice, (float) width, (float) height, dpi);


            using (var drawingSession = imageSource.CreateDrawingSession(Colors.White))
            {
                try
                {
                    var pixels = await backgroundBitmap.GetPixelsAsync();
                    var bitmap = SoftwareBitmap.CreateCopyFromBuffer(pixels, BitmapPixelFormat.Bgra8,
                        backgroundBitmap.PixelWidth, backgroundBitmap.PixelHeight);
                    var convertedImage = SoftwareBitmap.Convert(
                        bitmap,
                        BitmapPixelFormat.Bgra8,
                        BitmapAlphaMode.Premultiplied
                    );
                    var background = CanvasBitmap.CreateFromSoftwareBitmap(_canvasDevice, convertedImage);
                    drawingSession.DrawImage(background,new Rect(0,0,
                        width, 
                        height));
                    drawingSession.DrawInk(inkStrokes);
           
                    return imageSource;
                }
                catch (Exception e) when (_canvasDevice.IsDeviceLost(e.HResult))
                {
                    _canvasDevice.RaiseDeviceLost();
                }

            return null;

            }
        }


        public async Task<SoftwareBitmap> RenderToBitmapAsync(
            RenderTargetBitmap renderTargetBitmap,
            IReadOnlyList<InkStroke> inkStrokes,
            IReadOnlyList<InkStroke> highlightStrokes, double width, double height)
        {

            var dpi = DisplayInformation.GetForCurrentView().LogicalDpi;
            var renderTarget = new CanvasRenderTarget(_canvasDevice, (float)width, (float)height, dpi);
            using (renderTarget)
            {
                try
                {
                    using (var drawingSession = renderTarget.CreateDrawingSession())
                    {
                        var pixels = await renderTargetBitmap.GetPixelsAsync();
                        var bitmap = SoftwareBitmap.CreateCopyFromBuffer(pixels, BitmapPixelFormat.Bgra8, 
                            renderTargetBitmap.PixelWidth, renderTargetBitmap.PixelHeight);
                        var convertedImage = SoftwareBitmap.Convert(
                            bitmap,
                            BitmapPixelFormat.Bgra8,
                            BitmapAlphaMode.Premultiplied
                        );
                        var background = CanvasBitmap.CreateFromSoftwareBitmap(_canvasDevice, convertedImage);
                        drawingSession.DrawImage(background, new Rect(0,0,
                            width, height));
                        drawingSession.DrawInk(inkStrokes);
                    }

                    return await SoftwareBitmap.CreateCopyFromSurfaceAsync(renderTarget, BitmapAlphaMode.Premultiplied);
                }
                catch (Exception e) when (_canvasDevice.IsDeviceLost(e.HResult))
                {
                    _canvasDevice.RaiseDeviceLost();
                }

            }
            return null;
        }
    
        public void Trim()
        {
            _canvasDevice.Trim();
        }
       
        private void HandleDeviceLost(CanvasDevice sender, object args)
        {
            if (sender == _canvasDevice)
            {
                RecreateDevice();
            }
        }

        private void RecreateDevice()
        {
            _canvasDevice.DeviceLost -= HandleDeviceLost;

            _canvasDevice = CanvasDevice.GetSharedDevice(_canvasDevice.ForceSoftwareRenderer);
            _canvasDevice.DeviceLost += HandleDeviceLost;
        }
    }
}
